# Thesis Experiment: t5 Model
## Michael LeVine, April 19, 2024
## Hyperparameter testing: batch size 32

The purpose of this notebook is to test the summarization capabilities of the t5 mode when fine-tuning it with a batch size of 32l.

Attribution: This approach is based on a training course of Janana Ravi, Certified Google Cloud Architect and Data Engineer - from the LinkedIn Learning course: AI Text Summarization with Hugging Face, released 10/30/2023.

## Overview: Using a Transformer Model from Hugging Face: t5

### The t5 model
The pre-trained model that we will use is the "t5-small" model from Hugging Face, which can be found here: https://huggingface.co/google-t5/t5-small

The "model card," which describes the model, notes that it is a "text-to-text" transformer model.  The model card notes that t5 base is the model checkpoint with 60 million parameters

## Verifying the Compute Environment

### Graphics Processing Unit (GPU)
Running inference on transformer models can be done without a GPU.  However, for training, a GPU is recommended.  The following block of code shows checks whether a GPU is available for use in a PyTorch environment. 
.

In [1]:
import torch

if torch.cuda.is_available():
    device = torch.device("cuda")
else:
    device = torch.device("cpu")
    
print("using", device, "device")

using cuda device


## Installing and importing required libraries and dependencies

In [2]:
#command line pip install the necessary required libraries and dependencies
#the transformers library allows us to access the pre-trained t5-small model
#the datasets library provides access to the Hugging Face datasets
#the evaluate model enables us to evaluate the summarizations the model produces
#the rouge_score is a standard evaluation metric used in text summarization tasks
#the accelerate function allows for distributed training on GPUs
#pip install transformers datasets evaluate rouge_score accelerate

### Import the transformers library

In [2]:
#this code imports needed libraries

import transformers
import datasets
import evaluate
import rouge_score
import accelerate

print(transformers.__version__) # verifies the transformers version

4.32.1


## Importing, Reducing, and Exploring the dataset

The experiment will use the CNN/Daily Mail dataset.  Two datasets will be created:
* Training Dataset
* Holdout Dataset (for running inference on the model)


### Instantiating a training dataset

In [ ]:
data_directory = '/home/anthony/Michael LeVine/data'
file_name = 'your_data_file.csv'  # Replace with your actual file name

In [3]:
#loading the dataset which was previously saved
from datasets import load_dataset
cnn_news_summary_ds = load_dataset("arrow", data_files={'train': 'data/cnn_news_summary_ds/train/data-00000-of-00001.arrow', 'test': 'data/cnn_news_summary_ds/test/data-00000-of-00001.arrow'})


Generating train split: 0 examples [00:00, ? examples/s]

Generating test split: 0 examples [00:00, ? examples/s]

In [4]:
cnn_news_summary_ds

DatasetDict({
    train: Dataset({
        features: ['article', 'highlights', 'id'],
        num_rows: 2296
    })
    test: Dataset({
        features: ['article', 'highlights', 'id'],
        num_rows: 575
    })
})

As per above output, now the dataset is broken down into two components:
* a `train` (training) dataset of 2296 articles
* a `test` dataset of 575



### Instantiating a holdout dataset (200 records)

The holdout dataset is used to run inference on both the "off-the-shelf" model and the fine-tuned model.  The purpose of having a holdout set is so the model is running inference on a different dataset from what it was trained on in order to test its performance.

In [5]:
#load holdout set for inference from a local csv (to ensure same order)
cnn_holdout_ds = load_dataset ("csv", data_files='data/cnn_holdout_ds.csv', split = "train[0:200]")
cnn_holdout_ds

Generating train split: 0 examples [00:00, ? examples/s]

/opt/conda/envs/gpu/lib/python3.11/site-packages/pyarrow/pandas_compat.py:373: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if _pandas_api.is_sparse(col):


Dataset({
    features: ['article', 'highlights', 'id'],
    num_rows: 200
})

### Exploring the Dataset

In [6]:
#dataset shape
cnn_news_summary_ds.shape

{'train': (2296, 3), 'test': (575, 3)}

The above output shows the cnn_dailymail `train` set is 2296 rows x 3 columns.  The `test` set is 575 rows x 3 columns.

In [7]:
#dataset object type
type(cnn_news_summary_ds)

datasets.dataset_dict.DatasetDict

The above output shows the cnn_dailymail dataset is of type Dataset, within the datasets library.

In [8]:
#dataset structure
cnn_news_summary_ds

DatasetDict({
    train: Dataset({
        features: ['article', 'highlights', 'id'],
        num_rows: 2296
    })
    test: Dataset({
        features: ['article', 'highlights', 'id'],
        num_rows: 575
    })
})

The Dataset has three features: 
* `article`: The full text of the news article
*  `highlights`: the target summary, also known as the reference summary
*  `id`: the unique id for each article/highlights pair


In [9]:
#looking at the features of the dataset
cnn_news_summary_ds['train'].features

{'article': Value(dtype='string', id=None),
 'highlights': Value(dtype='string', id=None),
 'id': Value(dtype='string', id=None)}

In [10]:
#examining the first record of the dataset.  
cnn_news_summary_ds['train'][0]

{'article': '(CNN) -- After almost 10 months, the FBI has zeroed in on a suspect in the case of missing Florida pilot Robert Wiles, who may have been kidnapped for ransom. Missing Florida pilot Robert Wiles is thought to have been kidnapped for ransom. "We\'re close to solving the case," said FBI special agent David Couvertier. He would not elaborate. Agents also would not identify the suspect, and they said the person is not in custody. Investigators would only reveal that the "key suspect" is in Florida, either in Orlando, Lakeland or Melbourne. "They\'re holding that back in hopes of getting additional information," said Couvertier. The FBI says it\'s also looking at several persons of interest in those same three Florida cities. Wiles, 27, was last seen in the family\'s aircraft maintenance business, National Flight Services, at Lakeland Linder Regional Airport on April 1, 2008. The day Wiles disappeared, he left behind his bags, his computer, and even his car. His father says the 

## Preprocessing and Cleaning the Data

In [11]:
#defining a text cleaning function.  This function iterates over the 'article'
# and 'highlights' section and replaces various text strings (like
#backslashes, new lines, etc.) with the empty string

def clean_txt(example):
  for txt in ['article', 'highlights']:
    example[txt] = example[txt].lower() #convert text to lowercase
    example[txt] = example[txt].replace('\\','')
    example[txt] = example[txt].replace('/','')
    example[txt] = example[txt].replace('\n','')
    example[txt] = example[txt].replace('``','')
    example[txt] = example[txt].replace('"','')
    example[txt] = example[txt].replace('--','')
  return example

### Mapping the text cleaning function to the training dataset and the holdout dataset

In [12]:
#hugging face datasets allow the .map operation to
#apply a function to all records in a dataset, and then will
#update the dataset.  In efect, the .map() method
# maps the `clean_txt` function to
# all the records in the `cnn_news_summary_ds dataset.
#The result will be that the training and
#test data will now be cleaned
cleaned_cnn_news_summary_ds = cnn_news_summary_ds.map(clean_txt)

cleaned_cnn_news_summary_ds

Map:   0%|          | 0/2296 [00:00<?, ? examples/s]

Map:   0%|          | 0/575 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['article', 'highlights', 'id'],
        num_rows: 2296
    })
    test: Dataset({
        features: ['article', 'highlights', 'id'],
        num_rows: 575
    })
})

In [13]:
#verifying that we have a clean training dataset by comparing a record from the original
#dataset to the cleaned dataset
print('\n\n==== Original dataset ====\n\n')

print (cnn_news_summary_ds['train']['article'][0])

print('\n\n==== Cleaned dataset ====\n\n')

cleaned_cnn_news_summary_ds['train']['article'][0]



==== Original dataset ====


(CNN) -- After almost 10 months, the FBI has zeroed in on a suspect in the case of missing Florida pilot Robert Wiles, who may have been kidnapped for ransom. Missing Florida pilot Robert Wiles is thought to have been kidnapped for ransom. "We're close to solving the case," said FBI special agent David Couvertier. He would not elaborate. Agents also would not identify the suspect, and they said the person is not in custody. Investigators would only reveal that the "key suspect" is in Florida, either in Orlando, Lakeland or Melbourne. "They're holding that back in hopes of getting additional information," said Couvertier. The FBI says it's also looking at several persons of interest in those same three Florida cities. Wiles, 27, was last seen in the family's aircraft maintenance business, National Flight Services, at Lakeland Linder Regional Airport on April 1, 2008. The day Wiles disappeared, he left behind his bags, his computer, and even his car. His fa

"(cnn)  after almost 10 months, the fbi has zeroed in on a suspect in the case of missing florida pilot robert wiles, who may have been kidnapped for ransom. missing florida pilot robert wiles is thought to have been kidnapped for ransom. we're close to solving the case, said fbi special agent david couvertier. he would not elaborate. agents also would not identify the suspect, and they said the person is not in custody. investigators would only reveal that the key suspect is in florida, either in orlando, lakeland or melbourne. they're holding that back in hopes of getting additional information, said couvertier. the fbi says it's also looking at several persons of interest in those same three florida cities. wiles, 27, was last seen in the family's aircraft maintenance business, national flight services, at lakeland linder regional airport on april 1, 2008. the day wiles disappeared, he left behind his bags, his computer, and even his car. his father says the next day, wiles was supp

In [14]:
cleaned_cnn_holdout_ds = cnn_holdout_ds.map(clean_txt)

cleaned_cnn_holdout_ds

Map:   0%|          | 0/200 [00:00<?, ? examples/s]

Dataset({
    features: ['article', 'highlights', 'id'],
    num_rows: 200
})

In [15]:
#verifying that we have a clean holdout dataset by comparing a record from the original
#holdout dataset to the cleaned holdout dataset
print('\n\n==== Original holdout dataset ====\n\n')

print (cnn_holdout_ds['article'][0])

print('\n\n==== Cleaned holdout dataset ====\n\n')

cleaned_cnn_holdout_ds['article'][0]



==== Original holdout dataset ====


Manchester United have fallen off their perch. And they’re dropping like a stone towards mediocrity. That is the undeniable fact that has been hammered home relentlessly during the past six months. Whether we are talking about the events of Wednesday night at Olympiacos or before the startled eyes of the faithful at Old Trafford, the evidence is there for all to see. Can't stop the slump: David Moyes cannot believe it as he watches Manchester United lose at Olympiacos . Down and almost out: Robin van Persie lies on the floor during a defeat which sees United's Champions League campaign hanging by a thread . Disbelief: Wayne Rooney cries out in vain during another shambolic United display . Abject: The frustration shows on the Man United players' faces on taking the restart after conceding to Olympiacos . Coming to get you: Liverpool are looking to take United's place in the top four . Now is it time for Man United to sack Moyes? Out of the title r

"manchester united have fallen off their perch. and they’re dropping like a stone towards mediocrity. that is the undeniable fact that has been hammered home relentlessly during the past six months. whether we are talking about the events of wednesday night at olympiacos or before the startled eyes of the faithful at old trafford, the evidence is there for all to see. can't stop the slump: david moyes cannot believe it as he watches manchester united lose at olympiacos . down and almost out: robin van persie lies on the floor during a defeat which sees united's champions league campaign hanging by a thread . disbelief: wayne rooney cries out in vain during another shambolic united display . abject: the frustration shows on the man united players' faces on taking the restart after conceding to olympiacos . coming to get you: liverpool are looking to take united's place in the top four . now is it time for man united to sack moyes? out of the title race, out of the fa . cup, out of the l

In [16]:
cleaned_cnn_holdout_ds.to_csv('data/cleaned_cnn_holdout_ds.csv')

Creating CSV from Arrow format:   0%|          | 0/1 [00:00<?, ?ba/s]

922232

In [17]:
prefix = "summarize: "


def preprocess_function(examples):
    inputs = [prefix + doc for doc in examples["article"]]
    model_inputs = tokenizer(inputs, max_length=1024, truncation=True)

    labels = tokenizer(text_target=examples["highlights"], max_length=256, truncation=True)

    model_inputs["labels"] = labels["input_ids"]
    return model_inputs

## Instantiating an "off-the-shelf" t5 model and tokenizer

In [18]:
from transformers import pipeline, T5ForConditionalGeneration, TrainingArguments, Trainer, \
                         DataCollatorForSeq2Seq, T5Tokenizer
import pandas as pd
from datasets import Dataset
import random

In [19]:
from transformers import AutoTokenizer

checkpoint = "google-t5/t5-small"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

In [20]:
#looking at model architecture
from transformers import AutoModelForSeq2SeqLM, Seq2SeqTrainingArguments, Seq2SeqTrainer

model = AutoModelForSeq2SeqLM.from_pretrained(checkpoint)
model

T5ForConditionalGeneration(
  (shared): Embedding(32128, 512)
  (encoder): T5Stack(
    (embed_tokens): Embedding(32128, 512)
    (block): ModuleList(
      (0): T5Block(
        (layer): ModuleList(
          (0): T5LayerSelfAttention(
            (SelfAttention): T5Attention(
              (q): Linear(in_features=512, out_features=512, bias=False)
              (k): Linear(in_features=512, out_features=512, bias=False)
              (v): Linear(in_features=512, out_features=512, bias=False)
              (o): Linear(in_features=512, out_features=512, bias=False)
              (relative_attention_bias): Embedding(32, 8)
            )
            (layer_norm): T5LayerNorm()
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (1): T5LayerFF(
            (DenseReluDense): T5DenseActDense(
              (wi): Linear(in_features=512, out_features=2048, bias=False)
              (wo): Linear(in_features=2048, out_features=512, bias=False)
              (dropout): Drop

## Preprocessing Datasets

In [21]:
#define a preprocessing function

prefix = "summarize: "


def preprocess_function(examples):
    inputs = [prefix + doc for doc in examples["article"]]
    model_inputs = tokenizer(inputs, max_length=1024, truncation=True)

    labels = tokenizer(text_target=examples["highlights"], max_length=256, truncation=True)

    model_inputs["labels"] = labels["input_ids"]
    return model_inputs

### Preprocessing the training dataset

In [22]:
tokenized_cnn_training_ds = cleaned_cnn_news_summary_ds.map(preprocess_function, batched=True)
tokenized_cnn_training_ds

Map:   0%|          | 0/2296 [00:00<?, ? examples/s]

Map:   0%|          | 0/575 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['article', 'highlights', 'id', 'input_ids', 'attention_mask', 'labels'],
        num_rows: 2296
    })
    test: Dataset({
        features: ['article', 'highlights', 'id', 'input_ids', 'attention_mask', 'labels'],
        num_rows: 575
    })
})

### Preprocessing the holdout dataset

In [23]:
tokenized_cnn_holdout_ds = cleaned_cnn_holdout_ds.map(preprocess_function, batched=True)
tokenized_cnn_holdout_ds

Map:   0%|          | 0/200 [00:00<?, ? examples/s]

Dataset({
    features: ['article', 'highlights', 'id', 'input_ids', 'attention_mask', 'labels'],
    num_rows: 200
})

In [32]:
#tokenized_cnn_holdout_ds.to_csv('data/tokenized_cnn_holdout_ds.csv')

Creating CSV from Arrow format:   0%|          | 0/1 [00:00<?, ?ba/s]

1563405

## Instantiating a Data Collator

In [24]:
from transformers import DataCollatorForSeq2Seq

data_collator = DataCollatorForSeq2Seq(tokenizer=tokenizer, model=checkpoint)

## Importing ROUGE Metric
*  Rouge is a standard evaluation metric used in text summarization tasks
*  ROUGE provides an *objective metric* to compare model-produced summary with the dataset's reference summary.  

### Importing the evaluate library

The 'evaluate' library from Hugging Face allows us to evaluate ML models.  The 'evaluate' library provides access to dozens of evaluation metrics across many ML domains (including NLP, computer vision, etc.).

In [25]:
import evaluate

rouge = evaluate.load("rouge")
rouge

EvaluationModule(name: "rouge", module_type: "metric", features: [{'predictions': Value(dtype='string', id='sequence'), 'references': Sequence(feature=Value(dtype='string', id='sequence'), length=-1, id=None)}, {'predictions': Value(dtype='string', id='sequence'), 'references': Value(dtype='string', id='sequence')}], usage: """
Calculates average rouge scores for a list of hypotheses and references
Args:
    predictions: list of predictions to score. Each prediction
        should be a string with tokens separated by spaces.
    references: list of reference for each prediction. Each
        reference should be a string with tokens separated by spaces.
    rouge_types: A list of rouge types to calculate.
        Valid names:
        `"rouge{n}"` (e.g. `"rouge1"`, `"rouge2"`) where: {n} is the n-gram based scoring,
        `"rougeL"`: Longest common subsequence based scoring.
        `"rougeLsum"`: rougeLsum splits text using `"
"`.
        See details in https://github.com/huggingface/

In [43]:
#aggregated results of inference on holdout set
result_off_the_shelf_agg = rouge.compute(predictions = holdout_off_the_shelf_summaries,
                           references = holdout_article_summaries[:],
                           use_stemmer=True)

result_off_the_shelf_agg

{'rouge1': 0.37163589198592073,
 'rouge2': 0.1659438181778043,
 'rougeL': 0.26232053780080417,
 'rougeLsum': 0.2625393886964161}

In [44]:
#unaggregated results of inference on holdout set
result_off_the_shelf_unagg = rouge.compute(predictions = holdout_off_the_shelf_summaries,
                           references = holdout_article_summaries[:],
                           use_stemmer=True,
                             use_aggregator=False)

result_off_the_shelf_unagg

{'rouge1': [0.2839506172839506,
  0.45714285714285713,
  0.4054054054054054,
  0.352112676056338,
  0.3859649122807018,
  0.2,
  0.5393258426966292,
  0.3333333333333333,
  0.3684210526315789,
  0.2823529411764706,
  0.41666666666666663,
  0.46031746031746035,
  0.40449438202247195,
  0.3392857142857143,
  0.5057471264367817,
  0.29729729729729726,
  0.25396825396825395,
  0.35897435897435903,
  0.6233766233766234,
  0.43243243243243246,
  0.42105263157894735,
  0.6923076923076923,
  0.28571428571428575,
  0.4155844155844156,
  0.15384615384615385,
  0.24242424242424243,
  0.3137254901960784,
  0.33043478260869563,
  0.3555555555555555,
  0.5569620253164557,
  0.34285714285714286,
  0.32,
  0.3826086956521739,
  0.3103448275862069,
  0.5185185185185185,
  0.2975206611570248,
  0.4000000000000001,
  0.43809523809523804,
  0.4225352112676056,
  0.3636363636363636,
  0.4639999999999999,
  0.54,
  0.4878048780487805,
  0.31111111111111117,
  0.3157894736842105,
  0.3013698630136986,
  0.32

## Fine-tuning the t5-small model

In [26]:
#defining a custom compute metrics function
import numpy as np


def compute_metrics(eval_pred):
    predictions, labels = eval_pred
    decoded_preds = tokenizer.batch_decode(predictions, skip_special_tokens=True)
    labels = np.where(labels != -100, labels, tokenizer.pad_token_id)
    decoded_labels = tokenizer.batch_decode(labels, skip_special_tokens=True)

    result = rouge.compute(predictions=decoded_preds, references=decoded_labels, use_stemmer=True)

    prediction_lens = [np.count_nonzero(pred != tokenizer.pad_token_id) for pred in predictions]
    result["gen_len"] = np.mean(prediction_lens)

    return {k: round(v, 4) for k, v in result.items()}

In [27]:
training_args = Seq2SeqTrainingArguments(
    output_dir="t5_small_results",
    evaluation_strategy="epoch",
    learning_rate=2e-5,
    per_device_train_batch_size=32,#changed batch size from 16 to 32
    per_device_eval_batch_size=32, #changed batch size from 16 to 32
    weight_decay=0.01,
    save_total_limit=3,
    num_train_epochs=4,
    predict_with_generate=True,
    fp16=True,
    push_to_hub=False,
    load_best_model_at_end=True, #even if we overtrain model by accident, we will still 
    #load the checkpoint that had lowest evaluation loss

    #evaluation_strategy can be steps or epochs - correlates to how often we stop training and evaluate our model
    eval_steps=50,
    save_strategy='epoch' #save the model after every epoch
)

trainer = Seq2SeqTrainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_cnn_training_ds["train"],
    eval_dataset=tokenized_cnn_training_ds["test"],
    tokenizer=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics
    
)
trainer.evaluate() #adding a max_new_tokens paramater)

/opt/conda/envs/gpu/lib/python3.11/site-packages/accelerate/accelerator.py:432: FutureWarning: Passing the following arguments to `Accelerator` is deprecated and will be removed in version 1.0 of Accelerate: dict_keys(['dispatch_batches']). Please pass an `accelerate.DataLoaderConfiguration` instead: 
dataloader_config = DataLoaderConfiguration(dispatch_batches=None)
  warnings.warn(
You're using a T5TokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
/opt/conda/envs/gpu/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  warnings.warn('Was asked to gather along dimension 0, but all '


{'eval_loss': 2.0393264293670654,
 'eval_rouge1': 0.2204,
 'eval_rouge2': 0.0916,
 'eval_rougeL': 0.1819,
 'eval_rougeLsum': 0.1816,
 'eval_gen_len': 19.0,
 'eval_runtime': 32.7618,
 'eval_samples_per_second': 17.551,
 'eval_steps_per_second': 0.275}

In [29]:
training_args = Seq2SeqTrainingArguments(
    output_dir="t5_small_results",
    evaluation_strategy="epoch",
    learning_rate=2e-5,
    per_device_train_batch_size=32, #changed batch size from 16 to 32
    per_device_eval_batch_size=32,#changed batch size from 16 to 32
    weight_decay=0.01,
    save_total_limit=3,
    num_train_epochs=4,
    predict_with_generate=True,
    fp16=True,
    push_to_hub=False,
    load_best_model_at_end=True, #even if we overtrain model by accident, we will still 
    #load the checkpoint that had lowest evaluation loss

    #evaluation_strategy can be steps or epochs - correlates to how often we stop training and evaluate our model
    eval_steps=50,
    save_strategy='epoch' #save the model after every epoch
)

trainer = Seq2SeqTrainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_cnn_training_ds["train"],
    eval_dataset=tokenized_cnn_training_ds["test"],
    tokenizer=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics,
)

trainer.train() #adding a max_new_tokens paramater)

/opt/conda/envs/gpu/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  warnings.warn('Was asked to gather along dimension 0, but all '
/opt/conda/envs/gpu/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  warnings.warn('Was asked to gather along dimension 0, but all '
/opt/conda/envs/gpu/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  warnings.warn('Was asked to gather along dimension 0, but all '


TrainOutput(global_step=144, training_loss=1.8958524068196614, metrics={'train_runtime': 602.271, 'train_samples_per_second': 15.249, 'train_steps_per_second': 0.239, 'total_flos': 2485958209437696.0, 'train_loss': 1.8958524068196614, 'epoch': 4.0})

In [30]:
trainer.save_model()

## Instantiating the fine_tuned model and running inference with that model

In [31]:
fine_tuned_checkpoint = "t5_small_results"

In [32]:
#instantiating the fine-tuned model
fine_tuned_model = T5ForConditionalGeneration.from_pretrained(fine_tuned_checkpoint)

In [33]:
fine_tuned_model

T5ForConditionalGeneration(
  (shared): Embedding(32128, 512)
  (encoder): T5Stack(
    (embed_tokens): Embedding(32128, 512)
    (block): ModuleList(
      (0): T5Block(
        (layer): ModuleList(
          (0): T5LayerSelfAttention(
            (SelfAttention): T5Attention(
              (q): Linear(in_features=512, out_features=512, bias=False)
              (k): Linear(in_features=512, out_features=512, bias=False)
              (v): Linear(in_features=512, out_features=512, bias=False)
              (o): Linear(in_features=512, out_features=512, bias=False)
              (relative_attention_bias): Embedding(32, 8)
            )
            (layer_norm): T5LayerNorm()
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (1): T5LayerFF(
            (DenseReluDense): T5DenseActDense(
              (wi): Linear(in_features=512, out_features=2048, bias=False)
              (wo): Linear(in_features=2048, out_features=512, bias=False)
              (dropout): Drop

In [34]:
#instantiating the fine-tuned tokeinizer
tokenizer = AutoTokenizer.from_pretrained(fine_tuned_checkpoint)

### Running inference on holdout set with "fine-tuned" t5-small model

In [35]:
#instantiating a summarizer pipeline with the fine-tuned t5-small model
fine_tuned_summarizer = pipeline ('summarization', model=fine_tuned_checkpoint)

In [ ]:
#creating 'holdout_article_texts' variable to hold articles from the test set
holdout_article_texts = tokenized_cnn_holdout_ds["article"]

#creating 'holdout_article_summaries' variable to hold summaries from the test set
holdout_article_summaries = tokenized_cnn_holdout_ds["highlights"]


In [38]:
#running inference on holdout set with fine-tuned
from tqdm import tqdm

#instantiating an empty list named 'holdout summaries'
holdout_fine_tuned_summaries = []

prefix ='summarize: '

for i, text in enumerate(tqdm(holdout_article_texts[:500])):

    candidate = fine_tuned_summarizer(prefix + text)

    holdout_fine_tuned_summaries.append(candidate[0]["summary_text"])

100%|█████████████████████████████████████████| 200/200 [04:50<00:00,  1.45s/it]


In [55]:
cleaned_cnn_holdout_df['t5_fine_tuned_summaries'] = holdout_fine_tuned_summaries
cleaned_cnn_holdout_df

article  \
0    manchester united have fallen off their perch....   
1    a mother whose russian husband snatched their ...   
2    claim: supporters of mayor lutfur rahman alleg...   
3    the 15-year-old cousin of a palestinian boy wh...   
4    it may have made its way up the pole to become...   
..                                                 ...   
195  reality tv show the block has been accused of ...   
196  the average cost of raising a child to seconda...   
197  thai police investigating the murder of two br...   
198  from clumpy flat shoes that seem to shorten a ...   
199  for most people, a trip to the shops involves ...   

                                            highlights  \
0    manchester united were beaten 2-0 in their cha...   
1    rachael neustadt's sons - daniel, eight and jo...   
2    islamic voters allegedly told to be 'good musl...   
3    mohammed abu khder, 16, abducted and burned to...   
4    spearmint rhino records £2.1m loss in 2011 .lo...   
..                                                 ...   
195  'the block' caught out faking a visit from the...   
196  average cost of raising a child from birth up ...   
197  pornprasit sukdam claims he was offered £13,30...   
198  clumpy flat shoes that seem to shorten a woman...   
199  moose spent seven hours dashing through the fi...   

                                           id  \
0    5b3a626078390cb0e05327b4019753fd11cb8cea   
1    59d478d4a4299e2192997e56a9db9003fa2bac2d   
2    ec961b7d0912e7753dffe4360b77481eba96f2e1   
3    092d90d61eb105b3955820cc4894ac2c4995ad1b   
4    d0d59018cdf48aaeb6e1838c0323f8555e800765   
..                                        ...   
195  985b1bf7fc710e4ffdd9dd02e72d889a7997e89d   
196  e466296e19d7a14cf4916d70a2cbc296e4659c99   
197  a07624a84fe59a3321e83f153d6fd615207a8545   
198  e16474b52bbf45f49434fc4a0b1d68e2d3fba3c3   
199  6f4bbcea20c6d6132b5d987912e06b5f6099a6ea   

                                          t5_summaries  \
0    manchester united lost 1-0 to olympiacos in th...   
1    rachael neustadt's sons daniel, eight, and jon...   
2    supporters of mayor lutfur rahman allegedly ha...   
3    mohammed abu khder, 16, was abducted from his ...   
4    spearmint rhino has recorded a loss of £2.1mil...   
..                                                 ...   
195  reality tv show the block has been accused of ...   
196  average cost of raising a child to secondary s...   
197  pornprasit sukdam claims he was offered 700,00...   
198  kim carillo says she feels surprisingly sexy i...   
199  runaway elk caught dashing through streets of ...   

                               t5_fine_tuned_summaries  
0    manchester united lost 1-0 to olympiacos in th...  
1    rachael neustadt's sons daniel, eight, and jon...  
2    supporters of mayor lutfur rahman allegedly ha...  
3    mohammed abu khder, 16, was burned to death in...  
4    spearmint rhino has recorded a loss of £2.1m i...  
..                                                 ...  
195  reality tv show the block has been accused of ...  
196  average cost of raising a child to secondary s...  
197  pornprasit sukdam claims he was offered 700,00...  
198  kim carillo, who usually favours a more alluri...  
199  runaway elk caught dashing through streets of ...  

[200 rows x 5 columns]

## Evaluating the fine-tuned t5 performance

In [39]:
#aggregated results of inference on holdout set using fine-tune dmodel
result_fine_tuned_agg = rouge.compute(predictions = holdout_fine_tuned_summaries,
                           references = holdout_article_summaries[:],
                           use_stemmer=True)

result_fine_tuned_agg

{'rouge1': 0.37165211524244746,
 'rouge2': 0.1724968660090168,
 'rougeL': 0.26523861858664466,
 'rougeLsum': 0.2650452138248767}

In [40]:
#unaggregated results of inference on holdout set using fine-tune dmodel
result_fine_tuned_unagg = rouge.compute(predictions = holdout_fine_tuned_summaries,
                           references = holdout_article_summaries[:],
                           use_stemmer=True,
                            use_aggregator=False)
                                      

result_fine_tuned_unagg

{'rouge1': [0.28187919463087246,
  0.36507936507936506,
  0.4307692307692308,
  0.4634146341463415,
  0.5769230769230769,
  0.23529411764705882,
  0.4269662921348315,
  0.3661971830985915,
  0.2769230769230769,
  0.3142857142857143,
  0.41975308641975306,
  0.46031746031746035,
  0.35,
  0.4948453608247423,
  0.425531914893617,
  0.2894736842105263,
  0.3333333333333333,
  0.4210526315789473,
  0.6233766233766234,
  0.3030303030303031,
  0.4301075268817204,
  0.7105263157894738,
  0.28571428571428575,
  0.4691358024691358,
  0.2702702702702703,
  0.2337662337662338,
  0.45161290322580644,
  0.4166666666666667,
  0.29411764705882354,
  0.6285714285714286,
  0.30303030303030304,
  0.3888888888888889,
  0.3965517241379311,
  0.21428571428571427,
  0.25974025974025977,
  0.2727272727272727,
  0.45528455284552843,
  0.4220183486238532,
  0.4117647058823529,
  0.4,
  0.6614173228346456,
  0.54,
  0.46913580246913583,
  0.3561643835616438,
  0.41269841269841273,
  0.3934426229508197,
  0.3820

In [64]:
cleaned_cnn_holdout_df.to_csv('data/data_frame_after_t5.csv')

In [65]:
test_df = pd.read_csv('data/data_frame_after_t5.csv')
test_df

Unnamed: 0                                            article  \
0             0  manchester united have fallen off their perch....   
1             1  a mother whose russian husband snatched their ...   
2             2  claim: supporters of mayor lutfur rahman alleg...   
3             3  the 15-year-old cousin of a palestinian boy wh...   
4             4  it may have made its way up the pole to become...   
..          ...                                                ...   
195         195  reality tv show the block has been accused of ...   
196         196  the average cost of raising a child to seconda...   
197         197  thai police investigating the murder of two br...   
198         198  from clumpy flat shoes that seem to shorten a ...   
199         199  for most people, a trip to the shops involves ...   

                                            highlights  \
0    manchester united were beaten 2-0 in their cha...   
1    rachael neustadt's sons - daniel, eight and jo...   
2    islamic voters allegedly told to be 'good musl...   
3    mohammed abu khder, 16, abducted and burned to...   
4    spearmint rhino records £2.1m loss in 2011 .lo...   
..                                                 ...   
195  'the block' caught out faking a visit from the...   
196  average cost of raising a child from birth up ...   
197  pornprasit sukdam claims he was offered £13,30...   
198  clumpy flat shoes that seem to shorten a woman...   
199  moose spent seven hours dashing through the fi...   

                                           id  \
0    5b3a626078390cb0e05327b4019753fd11cb8cea   
1    59d478d4a4299e2192997e56a9db9003fa2bac2d   
2    ec961b7d0912e7753dffe4360b77481eba96f2e1   
3    092d90d61eb105b3955820cc4894ac2c4995ad1b   
4    d0d59018cdf48aaeb6e1838c0323f8555e800765   
..                                        ...   
195  985b1bf7fc710e4ffdd9dd02e72d889a7997e89d   
196  e466296e19d7a14cf4916d70a2cbc296e4659c99   
197  a07624a84fe59a3321e83f153d6fd615207a8545   
198  e16474b52bbf45f49434fc4a0b1d68e2d3fba3c3   
199  6f4bbcea20c6d6132b5d987912e06b5f6099a6ea   

                                          t5_summaries  \
0    manchester united lost 1-0 to olympiacos in th...   
1    rachael neustadt's sons daniel, eight, and jon...   
2    supporters of mayor lutfur rahman allegedly ha...   
3    mohammed abu khder, 16, was abducted from his ...   
4    spearmint rhino has recorded a loss of £2.1mil...   
..                                                 ...   
195  reality tv show the block has been accused of ...   
196  average cost of raising a child to secondary s...   
197  pornprasit sukdam claims he was offered 700,00...   
198  kim carillo says she feels surprisingly sexy i...   
199  runaway elk caught dashing through streets of ...   

                               t5_fine_tuned_summaries  
0    manchester united lost 1-0 to olympiacos in th...  
1    rachael neustadt's sons daniel, eight, and jon...  
2    supporters of mayor lutfur rahman allegedly ha...  
3    mohammed abu khder, 16, was burned to death in...  
4    spearmint rhino has recorded a loss of £2.1m i...  
..                                                 ...  
195  reality tv show the block has been accused of ...  
196  average cost of raising a child to secondary s...  
197  pornprasit sukdam claims he was offered 700,00...  
198  kim carillo, who usually favours a more alluri...  
199  runaway elk caught dashing through streets of ...  

[200 rows x 6 columns]